torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp

```diff
@@ -0,0 +1,262 @@
+    std::shared_ptr<Graph>& graph,
+    std::vector<int64_t> shapes,
+    double scale,
+    int64_t zero_point) {
+  Node* cast_node = graph->create(Symbol::caffe2("Int8GivenTensorFill"));
+  cast_node->is_(Symbol::attr("shape"), shapes);
+  cast_node->i_(Symbol::attr("Y_zero_point"), zero_point);
+  cast_node->f_(Symbol::attr("Y_scale"), scale);
+  cast_node->s_(Symbol::attr("values"), data);
+  return cast_node;
+}
+
+Node* CreateQuantizedBias(
+    std::vector<int64_t> data,
+    std::shared_ptr<Graph>& graph,
+    std::vector<int64_t> shapes,
+    double scale,
+    int64_t zero_point) {
+  Node* cast_node = graph->create(Symbol::caffe2("Int8GivenIntTensorFill"));
+  cast_node->is_(Symbol::attr("shape"), shapes);
+  cast_node->i_(Symbol::attr("Y_zero_point"), zero_point);
+  cast_node->f_(Symbol::attr("Y_scale"), scale);
+  cast_node->is_(Symbol::attr("values"), data);
+  return cast_node;
+}
+
+// This is called before the onnx pass. Using pattern matching we
+// find the relevant nodes and extract the packed_params. The packed_params are
+// passed to the appropriate unpack function using c10::Dispatcher. We insert
+// the unpacked weights and bias into the graph using
+// caffe2::Int8GivenTensorFill nodes.
+void unpackQuantizedWeightsHelper(
+    std::shared_ptr<Graph>& graph,
+    std::map<std::string, at::Tensor>& paramsDict,
+    const std::string& pattern,
+    const std::string& unpack_fn) {
+  Graph pattern_graph;
+  std::unordered_map<std::string, Value*> vmap;
+  script::parseIR(pattern, &pattern_graph, vmap);
+  const auto& matches = findPatternMatches(pattern_graph, *graph);
+
+  for (const auto& match : matches) {
+    auto match_vmap = match.values_map;
+    auto qlinear_node = match_vmap.at(vmap.at("r"))->node();
+    std::string quantized_weight =
+        match_vmap.at(vmap.at("r"))->node()->inputs()[1]->debugName();
+
+    auto itr = paramsDict.find(quantized_weight);
+    if (itr == paramsDict.end()) {
+      throw std::runtime_error(
+          "getValues: Quantized weight value not found amongst constant parameters.");
+    }
+    at::Tensor packed_weight = itr->second;
+    auto op = Dispatcher::singleton().findSchema({unpack_fn, ""});
+    assert(op.has_value());
+    std::tuple<at::Tensor, c10::optional<at::Tensor>> result = callOpUnboxed<
+        std::tuple<at::Tensor, c10::optional<at::Tensor>>,
+        at::Tensor>(*op, packed_weight);
+    at::Tensor unpacked_weight = std::get<0>(result);
+
+    // Permute weights?
+    /*
+    if (unpacked_weight.ndimension() == 2) {
+      std::cout << "2 dim weight ... permuting \n";
+      unpacked_weight.permute({1, 0});
+    } else if (unpacked_weight.ndimension() == 4) {
+      std::cout << "4 dim weight ... permuting \n";
+      unpacked_weight.permute({0, 2, 3, 1});
+    }
+    */
+
+    // Remove packed_params
+    qlinear_node->removeInput(1);
+
+    // Convert from int8 to uint8
+    int8_t* inp_data = (int8_t*)unpacked_weight.data_ptr<c10::qint8>();
```

Comment:
i am a bit confused here. c10::qint8 is uint8? why do we need to do the conversion here? can we move the conversion to the actually assignment?

Reply:
Pytorch uses int8 for the weights. But caffe2 (qnnpack) expects uint8 values. Hence the conversion. We do the same for quantization on mobile as well where we use qnnpack.

Reply:
my question is what's the relationship between qint8 and int8_t?

Nit: reinterpret_cast<int8*>

Reply:
c10::qint8 is actually int8. I need to do this cast here so that I can convert to uint8 by adding 128 below and casting to char.